homeassistant/helpers/state.py

```diff
@@ -53,36 +70,67 @@ def reproduce_state(hass, states, blocking=False):
 
     to_call = defaultdict(list)
 
+    # Update this dict of lists when new services are added to HA.
+    # Each item is a service with a list of required attributes.
+    service_attributes = {
+        SERVICE_PLAY_MEDIA: [ATTR_MEDIA_CONTENT_TYPE, ATTR_MEDIA_CONTENT_ID],
+        SERVICE_MEDIA_SEEK: [ATTR_MEDIA_SEEK_POSITION],
+        SERVICE_VOLUME_MUTE: [ATTR_MEDIA_VOLUME_MUTED],
+        SERVICE_VOLUME_SET: [ATTR_MEDIA_VOLUME_LEVEL],
+        SERVICE_NOTIFY: [ATTR_MESSAGE],
+        SERVICE_SET_AWAY_MODE: [ATTR_AWAY_MODE],
+        SERVICE_SET_FAN_MODE: [ATTR_FAN],
+        SERVICE_SET_TEMPERATURE: [ATTR_TEMPERATURE],
+    }
+
+    # Update this dict when new services are added to HA.
+    # Each item is a service with a corresponding state.
+    service_to_state = {
+        SERVICE_TURN_ON: STATE_ON,
+        SERVICE_TURN_OFF: STATE_OFF,
+        SERVICE_MEDIA_PLAY: STATE_PLAYING,
+        SERVICE_MEDIA_PAUSE: STATE_PAUSED,
+        SERVICE_ALARM_ARM_AWAY: STATE_ALARM_ARMED_AWAY,
+        SERVICE_ALARM_ARM_HOME: STATE_ALARM_ARMED_HOME,
+        SERVICE_ALARM_DISARM: STATE_ALARM_DISARMED,
+        SERVICE_ALARM_TRIGGER: STATE_ALARM_TRIGGERED,
+        SERVICE_LOCK: STATE_LOCKED,
+        SERVICE_UNLOCK: STATE_UNLOCKED,
+        SERVICE_CLOSE: STATE_CLOSED,
+        SERVICE_OPEN: STATE_OPEN,
+        SERVICE_MOVE_UP: STATE_OPEN,
+        SERVICE_MOVE_DOWN: STATE_CLOSED,
+    }
+
     for state in states:
-        current_state = hass.states.get(state.entity_id)
 
-        if current_state is None:
+        if hass.states.get(state.entity_id) is None:
             _LOGGER.warning('reproduce_state: Unable to find entity %s',
                             state.entity_id)
             continue
 
-        if state.domain == 'media_player' and state.attributes and \
-            'media_type' in state.attributes and \
-                'media_id' in state.attributes:
-            service = SERVICE_PLAY_MEDIA
-        elif state.domain == 'media_player' and state.state == STATE_PAUSED:
-            service = SERVICE_MEDIA_PAUSE
-        elif state.domain == 'media_player' and state.state == STATE_PLAYING:
-            service = SERVICE_MEDIA_PLAY
-        elif state.state == STATE_ON:
-            service = SERVICE_TURN_ON
-        elif state.state == STATE_OFF:
-            service = SERVICE_TURN_OFF
+        if state.domain == GROUP_DOMAIN:
+            service_domain = HASS_DOMAIN
         else:
+            service_domain = state.domain
+
+        domain_services = hass.services.services[service_domain]
+
+        service = None
+        for _service in domain_services.keys():
+            if (_service in service_attributes and
+                    all(attr in state.attributes
+                        for attr in service_attributes[_service])):
+                service = _service
+            if (_service in service_to_state and
```

Comment:
Actually, I'm thinking I might have to duplicate the iteration over domain_services. Otherwise it will be unpredictable which service is selected if a user configures both a state and attribute(s) that target different services for one entity in the same yaml block.

I'd like to have state changes take precedence over attribute changes. Thoughts?